R/cellbrowser.R

```diff
@@ -0,0 +1,499 @@
+  markers.file = NULL,
+  cluster.field = NULL,
+  cb.dir = NULL,
+  port = NULL,
+  use.mtx = FALSE,
+  meta.fields = NULL,
+  meta.fields.names = NULL,
+  matrix.slot = "counts",
+  markers.n = 100,
+  skip.markers = FALSE,
+  skip.expr.matrix = FALSE,
+  skip.metadata = FALSE,
+  skip.reductions = FALSE
+) {
+  if (!requireNamespace("Seurat", quietly = TRUE)) {
+    stop("This script requires that Seurat (V2 or V3) is installed")
+  }
+  message("Seurat Version installed: ", packageVersion(pkg = "Seurat"))
+  object.version <- as.character(x = slot(object = object, name = 'version'))
+  message("Object was created with Seurat version ", object.version)
+  if (substr(x = object.version, start = 1, stop = 1) != '2' && substr(x = object.version, start = 1, stop = 1) != '3') {
+    stop("can only process Seurat2 or Seurat3 objects, version of rds is ", object.version)
+  }
+  if (substr(x = object.version, start = 1, stop = 1) != substr(x = packageVersion(pkg = "Seurat"), start = 1, stop = 1)) {
+    stop("The installed version of Seurat is different from Seurat object loaded. You have to down- or upgrade your installed Seurat version, see the Seurat documentation")
+  }
+  embeddings <- reductions # Seurat prefers the word reductions
+  # compatibility layer for Seurat 2 vs 3
+  # see https://satijalab.org/seurat/essential_commands.html
+  if (substr(x = object.version, start = 1, stop = 1) == '2') {
+    # Seurat 2 data access
+    idents <- object@ident # Idents() in Seurat3
+    meta <- object@meta.data
+    cellOrder <- object@cell.names
+    if (matrix.slot == "counts") {
+      counts <- object@raw.data
+    } else if (matrix.slot == "scale.data") {
+      counts <- object@scale.data
+    }
+    else if (matrix.slot == "data") {
+      counts <- object@data
+    } else {
+      stop("matrix.slot can only be one of: counts, scale.data, data")
+    }
+    genes <- rownames(x = object@data)
+    dr <- object@dr
+  } else {
+    # Seurat 3 functions
+    idents <- Idents(object = object)
+    meta <- object[[]]
+    cellOrder <- colnames(x = object)
+    counts <- findMatrix(object = object, matrix.slot = matrix.slot)
+    if (dim(x = counts)[1] == 0) {
+      message(paste0("The Seurat data slot '", matrix.slot, "' contains no data. Trying default assay."))
+      defAssay <- DefaultAssay(object)
+      assay <- GetAssay(object, defAssay)
+      message(paste0("Default assay is ", defAssay))
+      counts <- findMatrix(assay, matrix.slot)
+      genes <- rownames(counts)
+      if (dim(x = counts)[1] == 0) {
+        stop(
+          "Could not find an expression matrix",
+          "Please select the correct slot where the matrix is stored, possible ",
+          "values are 'counts', 'scale.data' or 'data'. To select a slot, ",
+          "use the option 'matrix.slot' from R or the cbImportSeurat option -s from the command line."
+        )
+      }
+    }
+    else {
+      genes <- rownames(x = object)
+    }
+    dr <- object@reductions
```

Comment:
Typically, we as that you use the accessors or `slot` to pull data from a Seurat v3 object instead of `@`